File: frontend/src/stores/alumni.js
import { defineStore } from "pinia";
import { ref } from "vue";
import alumniService from "@/services/alumni.service";
import { useToast } from "@/composables/useToast";

export const useAlumniStore = defineStore("alumni", () => {
  const toast = useToast();

  const items = ref([]);
  const pagination = ref({ page: 1, limit: 10, total: 0, total_pages: 1, has_next: false, has_previous: false });
  const current = ref(null);
  const candidates = ref([]);
  const loading = ref(false);
  const importing = ref(false);
  const searching = ref(false);
  const searchResults = ref([]);

  const fetchList = async (params = {}) => {
    loading.value = true;
    try {
      const res = await alumniService.list(params);
      items.value = res.data.data.alumni;
      pagination.value = res.data.data.pagination;
    } catch (err) {
      toast.error(err.response?.data?.detail || "Gagal memuat daftar alumni");
    } finally {
      loading.value = false;
    }
  };

  const fetchOne = async (id) => {
    loading.value = true;
    try {
      const res = await alumniService.get(id);
      current.value = res.data.data;
    } catch (err) {
      toast.error(err.response?.data?.detail || "Gagal memuat data alumni");
    } finally {
      loading.value = false;
    }
  };

  const fetchCandidates = async (id) => {
    try {
      const res = await alumniService.candidates(id);
      candidates.value = res.data.data;
    } catch (err) {
      toast.error(err.response?.data?.detail || "Gagal memuat kandidat");
    }
  };

  const create = async (payload) => {
    const res = await alumniService.create(payload);
    toast.success("Alumni berhasil ditambahkan");
    return res.data.data;
  };

  const update = async (id, payload) => {
    const res = await alumniService.update(id, payload);
    toast.success("Alumni berhasil diperbarui");
    return res.data.data;
  };

  const remove = async (id) => {
    await alumniService.remove(id);
    toast.success("Alumni berhasil dihapus");
  };

  const addManualCandidate = async (id, payload) => {
    const res = await alumniService.addManualCandidate(id, payload);
    toast.success("Temuan manual tersimpan");
    return res.data.data;
  };

  const searchWeb = async (id) => {
    searching.value = true;
    searchResults.value = [];
    try {
      const res = await alumniService.searchWeb(id);
      searchResults.value = res.data.data;
      return res.data.data;
    } catch (err) {
      toast.error(err.response?.data?.detail || "Pencarian web gagal");
      throw err;
    } finally {
      searching.value = false;
    }
  };

  const importExcel = async (file, options = {}) => {
    importing.value = true;
    try {
      const res = await alumniService.importExcel(file, options);
      const summary = res.data.data;
      if (!options.dryRun) {
        toast.success(
          `Impor selesai: ${summary.created} baru, ${summary.skipped_duplicate} duplikat dilewati` +
            (summary.skipped_invalid ? `, ${summary.skipped_invalid} baris tidak valid` : "")
        );
      }
      return summary;
    } catch (err) {
      toast.error(err.response?.data?.detail || "Impor gagal");
      throw err;
    } finally {
      importing.value = false;
    }
  };

  return {
    items,
    pagination,
    current,
    candidates,
    loading,
    importing,
    searching,
    searchResults,
    fetchList,
    fetchOne,
    fetchCandidates,
    create,
    update,
    remove,
    addManualCandidate,
    importExcel,
    searchWeb,
  };
});

File: backend/app/api/v1/alumni.py
from fastapi import APIRouter, Depends, File, HTTPException, Query, UploadFile, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.infrastructure.db import get_db
from app.infrastructure.models import User
from app.schemas.alumni import AlumniCreateRequest, AlumniImportResponse, AlumniResponse, AlumniUpdateRequest
from app.schemas.common import SuccessResponse, paginate
from app.services.alumni_service import AlumniService, to_alumni_response
from app.services.import_service import ImportError_, import_file
from app.services.import_enriched_service import import_enriched_file

router = APIRouter(prefix="/alumni", tags=["alumni"])

ALLOWED_IMPORT_EXTENSIONS = (".xlsx", ".xls", ".csv")
MAX_IMPORT_FILE_SIZE = 25 * 1024 * 1024  # 25 MB


@router.post("/import-enriched", response_model=SuccessResponse[dict])
async def import_enriched_alumni(
    file: UploadFile = File(...),
    dry_run: bool = Query(False, description="Hanya tampilkan ringkasan, jangan simpan ke DB"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Impor file hasil scraping OSINT (DailyProject4) langsung ke profil alumni & kandidat."""
    filename = file.filename or ""
    if not filename.lower().endswith(ALLOWED_IMPORT_EXTENSIONS):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Format file tidak didukung. Gunakan salah satu dari: {', '.join(ALLOWED_IMPORT_EXTENSIONS)}",
        )

    content = await file.read()
    if len(content) > MAX_IMPORT_FILE_SIZE:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Ukuran file melebihi batas 25 MB")

    try:
        summary = import_enriched_file(db, current_user.id, content, filename, dry_run=dry_run)
    except Exception as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Gagal impor: {exc}") from exc

    return SuccessResponse(
        message=f"Berhasil mengimpor {summary.candidates_created} temuan scraping!",
        data={
            "total_rows": summary.total_rows,
            "alumni_created": summary.alumni_created,
            "alumni_updated": summary.alumni_updated,
            "candidates_created": summary.candidates_created,
            "skipped_empty": summary.skipped_empty,
        }
    )


@router.post("/import", response_model=SuccessResponse[AlumniImportResponse])
async def import_alumni(
    file: UploadFile = File(...),
    dry_run: bool = Query(False, description="Hanya tampilkan ringkasan, jangan simpan ke DB"),
    limit: int | None = Query(None, ge=1, description="Batasi jumlah baris yang diimpor (uji coba)"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Impor data induk alumni (Nama Lulusan, NIM, Tahun Masuk, Tanggal Lulus, Fakultas,
    Program Studi) langsung dari file Excel/CSV roster kampus, ke akun yang sedang login.
    Baris dengan NIM yang sudah ada di akun ini dilewati (idempotent)."""
    filename = file.filename or ""
    if not filename.lower().endswith(ALLOWED_IMPORT_EXTENSIONS):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Format file tidak didukung. Gunakan salah satu dari: {', '.join(ALLOWED_IMPORT_EXTENSIONS)}",
        )

    content = await file.read()
    if len(content) > MAX_IMPORT_FILE_SIZE:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Ukuran file melebihi batas 25 MB")

    try:
        summary = import_file(db, current_user.id, content, filename, limit=limit, dry_run=dry_run)
    except ImportError_ as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    message = (
        f"Pratinjau: {summary.created} baru, {summary.skipped_duplicate} duplikat dilewati"
        if dry_run
        else f"Impor selesai: {summary.created} alumni baru ditambahkan"
    )
    return SuccessResponse(message=message, data=AlumniImportResponse(**summary.__dict__))


@router.post("", response_model=SuccessResponse[AlumniResponse], status_code=status.HTTP_201_CREATED)
def create_alumni(
    payload: AlumniCreateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    alumni = AlumniService(db).create(current_user.id, payload.model_dump())
    return SuccessResponse(message="Alumni berhasil ditambahkan", data=to_alumni_response(alumni))


@router.get("")
def list_alumni(
    page: int = Query(1, ge=1),
    limit: int = Query(10, ge=1, le=100),
    search: str | None = None,
    status_filter: str | None = Query(None, alias="status"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    items, total = AlumniService(db).list(current_user.id, page, limit, search, status_filter)
    return {
        "success": True,
        "message": "Daftar alumni",
        "data": {
            "alumni": [to_alumni_response(a) for a in items],
            "pagination": paginate(page, limit, total).model_dump(),
        },
    }


@router.get("/{alumni_id}", response_model=SuccessResponse[AlumniResponse])
def get_alumni(alumni_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    try:
        alumni = AlumniService(db).get_or_404(current_user.id, alumni_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return SuccessResponse(message="Detail alumni", data=to_alumni_response(alumni))


@router.put("/{alumni_id}", response_model=SuccessResponse[AlumniResponse])
def update_alumni(
    alumni_id: str,
    payload: AlumniUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        alumni = AlumniService(db).update(current_user.id, alumni_id, payload.model_dump(exclude_unset=True))
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return SuccessResponse(message="Alumni berhasil diperbarui", data=to_alumni_response(alumni))


@router.delete("/{alumni_id}", response_model=SuccessResponse[None])
def delete_alumni(alumni_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    try:
        AlumniService(db).delete(current_user.id, alumni_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return SuccessResponse(message="Alumni berhasil dihapus", data=None)
